src/TestResults/match-by-context.ts

```diff
@@ -0,0 +1,292 @@
+/**
+ * matching tests with assertion by its `context`, i.e. describe/test block structure
+ * as well as sequence (by relative line position) in its context.
+ *
+ * The assumption is both source parser and jest have generated correct output,
+ * while the names of the test might not always match (such as template-literal, jest.each use cases),
+ * nor the line numbers (due to source map difference and other transpile complication),
+ * the relative position should be the same, i.e. the first assertion in describe-block X
+ * should always match the first test block under the same describe block and so on.
+ */
+
+import { ItBlock, TestAssertionStatus, ParsedNode, DescribeBlock } from 'jest-editor-support';
+import { TestReconciliationState } from './TestReconciliationState';
+import { TestResult } from './TestResult';
+import { DataNode, ContainerNode, NodeType } from './MatchNode';
+
+type ContextType = 'container' | 'data';
+const ROOT_NODE_NAME = '__root__';
+export const buildAssertionContainer = (
+  assertions: TestAssertionStatus[]
+): ContainerNode<TestAssertionStatus> => {
+  const root = new ContainerNode<TestAssertionStatus>(ROOT_NODE_NAME);
+  if (assertions.length > 0) {
+    assertions.forEach((a) => {
+      const container = root.findContainer(a.ancestorTitles, true);
+      container.addDataNode(new DataNode(a.title, a.location?.line ?? 0, a));
+    });
+    // group by line since there could be multiple assertions for the same test block, such
+    // as in the jest.each use case
+    root.sort(true);
+  }
+  return root;
+};
+
+export const buildSourceContainer = (sourceRoot: ParsedNode): ContainerNode<ItBlock> => {
+  const isDescribeBlock = (node: ParsedNode): node is DescribeBlock => node.type === 'describe';
+  const isItBlock = (node: ParsedNode): node is ItBlock => node.type === 'it';
+  const buildNode = (node: ParsedNode, parent: ContainerNode<ItBlock>): void => {
+    let container = parent;
+    if (isDescribeBlock(node)) {
+      container = new ContainerNode(node.name);
+      parent.addContainerNode(container);
+    } else if (isItBlock(node)) {
+      parent.addDataNode(new DataNode(node.name, node.start.line - 1, node));
+    }
+
+    node.children?.forEach((n) => buildNode(n, container));
+  };
+
+  const root = new ContainerNode<ItBlock>(ROOT_NODE_NAME);
+  buildNode(sourceRoot, root);
+  // do not need to do grouping since there can't be test blocks that share lines
+  root.sort(false);
+  return root;
+};
+
+const matchPos = (t: ItBlock, a?: TestAssertionStatus, allowLocation = true): boolean => {
+  const line = a?.line ?? (allowLocation && a.location?.line);
+  return line >= t.start.line && line <= t.end.line;
```

Comment:
If `allowLocation` is false, `line` can evaluate to `false`, which in the process of comparison is converted to 0.
By the way, could you help me why we can't use `a.location` in some cases?